### PR TITLE
Spanner - add NUMERIC type

### LINKED ISSUE #2515: Add NUMERIC data type support for Spanner (BigDecimal)

### PR DESCRIPTION
**Pull request by dmitry-s**
fixes #2515

### CODE REVIEW

File: spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/convert/StructAccessor.java [hunk 1]
```diff
@@ -91,6 +94,7 @@
 				.put(double[].class, AbstractStructReader::getDoubleArray)
 				.put(long[].class, AbstractStructReader::getLongArray)
 				.put(boolean[].class, AbstractStructReader::getBooleanArray)
+				.put(BigDecimal.class, AbstractStructReader::getBigDecimal)
```
Comment (meltsufin) [inline spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/convert/StructAccessor.java:97]: This looks like a duplicate of the above map, but there's probably some obscure reason for it. :)

File: spring-cloud-gcp-data-spanner/src/test/java/org/springframework/cloud/gcp/data/spanner/repository/query/SpannerStatementQueryTests.java [hunk 2]
```diff
@@ -100,26 +102,28 @@ public void compoundNameConventionTest() throws NoSuchMethodException {
...
 
 					String expectedQuery =
-							"SELECT DISTINCT shares, trader_id, ticker, price, action, id "
+							"SELECT DISTINCT shares, trader_id, ticker, price, action, id, value "
 									+ "FROM trades WHERE ( LOWER(action)=LOWER(@tag0) "
 									+ "AND ticker=@tag1 ) OR "
 									+ "( trader_id=@tag2 AND price<@tag3 ) OR ( price>=@tag4 AND id<>NULL AND "
 									+ "trader_id=NULL AND trader_id LIKE @tag7 AND price=TRUE AND price=FALSE AND "
-									+ "price>@tag10 AND price<=@tag11 AND price IN UNNEST(@tag12) ) ORDER BY id DESC LIMIT 3";
+									+ "price>@tag10 AND price<=@tag11 AND price IN UNNEST(@tag12) "
```
Comment (meltsufin) [inline spring-cloud-gcp-data-spanner/src/test/java/org/springframework/cloud/gcp/data/spanner/repository/query/SpannerStatementQueryTests.java:125]: `price` is actually a perfect field to use `BigDecimal` for.

**Review by meltsufin — CHANGES_REQUESTED** [review]
Looks good overall!

Can you also update the sample where we use `Double price` to use `BigDecimal price`?

Looks like there is a compilation error in Firestore module now. Probably has to do with dep version changes.

**Review by elefeint — APPROVED** [review]
Couple of optional nits, LGTM.

File: spring-cloud-gcp-data-firestore/src/main/java/org/springframework/cloud/gcp/data/firestore/util/Util.java [hunk 3]
```diff
@@ -27,7 +25,20 @@ private Util() {
 	}
 
 	public static String extractDatabasePath(String parent) {
-		return parent.substring(0, StringUtils.ordinalIndexOf(parent, "/", 4));
+		//the parent looks like this: projects/{project_id}/databases/{database_id}/...
```
Comment (meltsufin) [inline spring-cloud-gcp-data-firestore/src/main/java/org/springframework/cloud/gcp/data/firestore/util/Util.java:28]: This is unrelated. Can we please put it in a separate PR?

**Review by meltsufin — CHANGES_REQUESTED** [review]
Looks good for Spanner, but can we do the Firestore fix in a separate PR please?

**Review by meltsufin — APPROVED** [review]

**Review by elefeint — APPROVED** [review]

**Pull request merged**